core/src/main/java/tech/pegasys/eth2signer/core/multikey/MultiKeyArtifactSignerProvider.java

```diff
@@ -12,72 +12,140 @@
+    if (signer.isEmpty()) {
+      LOG.error("No matching metadata file found for the identifier {}", signerIdentifier);
+      return Optional.empty();
+    }
+    final ArtifactSignerWithFileName signerWithFileName = signer.get();
+    if (!signerMatchesIdentifier(signerWithFileName, signerIdentifier)) {
+      LOG.error(
+          "Signing metadata file {} does not correspond to the specified signer identifier {}",
+          signerWithFileName.getPath().getFileName(),
+          signerWithFileName.getSigner().getIdentifier());
+      return Optional.empty();
+    }
+    return signer.map(ArtifactSignerWithFileName::getSigner);
   }
 
   @Override
-  public Set<String> availableSigners() {
-    return signingMetadataTomlConfigLoader.loadAvailableSigningMetadataTomlConfigs().stream()
-        .map(metadataFile -> metadataFile.createSigner(this))
+  public Set<String> availableIdentifiers() {
+    return loadAvailableSigners().stream()
         .filter(Objects::nonNull)
+        .map(ArtifactSignerWithFileName::getSigner)
         .map(ArtifactSigner::getIdentifier)
         .collect(Collectors.toSet());
   }
 
-  @Override
-  public ArtifactSigner createSigner(final UnencryptedKeyMetadataFile metadataFile) {
-    final ArtifactSigner signer;
-    signer = UnencryptedKeyFileSignerFactory.createSigner(metadataFile.getPrivateKeyBytes());
+  private Optional<ArtifactSignerWithFileName> loadSignerForIdentifier(
+      final String signerIdentifier) {
+    final Filter<Path> pathFilter = signerIdentifierFilenameFilter(signerIdentifier);
+    final Collection<ArtifactSignerWithFileName> matchingSigners = findSigners(pathFilter);
+    if (matchingSigners.size() > 1) {
+      LOG.error(
+          "Found multiple signing metadata file matches for signer identifier " + signerIdentifier);
+      return Optional.empty();
+    } else if (matchingSigners.isEmpty()) {
+      return Optional.empty();
+    } else {
+      return Optional.of(matchingSigners.iterator().next());
+    }
+  }
 
-    if (filenameMatchesSigningAddress(signer, metadataFile)) {
-      LOG.info("Loaded signer for address {}", signer.getIdentifier());
-      return signer;
+  private Collection<ArtifactSignerWithFileName> loadAvailableSigners() {
+    return findSigners((entry) -> true);
+  }
+
+  private Collection<ArtifactSignerWithFileName> findSigners(
+      final DirectoryStream.Filter<? super Path> filter) {
+    final Collection<ArtifactSignerWithFileName> signers = new HashSet<>();
+
+    try (final DirectoryStream<Path> directoryStream =
+        Files.newDirectoryStream(configsDirectory, filter)) {
+      for (final Path file : directoryStream) {
+        try {
+          final ArtifactSignerWithFileName artifactSignerWithFileName =
+              new ArtifactSignerWithFileName(file, signerParser.parse(file));
+          signers.add(artifactSignerWithFileName);
+        } catch (Exception e) {
+          LOG.error("Error parsing signing metadata file {}", file, e);
```

Comment:
nit: wondering - if you have 2 files which match the filter, but one of them fails - should you keep working? Or is the fact that more than 1 file matched a "fail" condition?

Reply:
I think you should keep working and attempt to load the next file